node/src/actors/chain_manager/mod.rs

```diff
@@ -721,12 +721,16 @@ impl ReputationInfo {
         self.alpha_diff += Alpha(u32::from(replication_factor));
 
         // Set of pkhs which were rewarded in the tally transaction
-        // TODO: Review case of data requester is also a witness
         let tally_vts: HashSet<_> = tally_transaction.outputs.iter().map(|vt| vt.pkh).collect();
+        let tally_liars: HashSet<_> = tally_transaction.out_of_consensus.iter().collect();
         for pkh in commits.keys() {
             // If the identity behaved honestly, it must have received a reward.
             // So we know which identities lied because they do not have a vtt transaction in the tally.
-            let liar = if tally_vts.contains(pkh) { 0 } else { 1 };
+            let liar = if tally_vts.contains(pkh) && !tally_liars.contains(pkh) {
```

Comment:
Isn't the first check redundant?

```suggestion
            let liar = if !tally_liars.contains(pkh) {
```

Reply:
non revealers has to be punished too

Reply:
I thought they were also inserted into `tally_liars`, doesn't that make more sense?